pkg/resources/database.go

```diff
@@ -2,395 +2,466 @@ package resources
+
+		CustomizeDiff: DatabaseParametersCustomDiff,
+		Schema:        MergeMaps(databaseSchema, DatabaseParametersSchema),
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
+
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Version: 0,
+				// setting type to cty.EmptyObject is a bit hacky here but following https://developer.hashicorp.com/terraform/plugin/framework/migrating/resources/state-upgrade#sdkv2-1 would require lots of repetitive code; this should work with cty.EmptyObject
+				Type:    cty.EmptyObject,
+				Upgrade: v092DatabaseStateUpgrader,
+			},
+		},
 	}
 }
 
-// CreateDatabase implements schema.CreateFunc.
-func CreateDatabase(d *schema.ResourceData, meta interface{}) error {
+func CreateDatabase(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
 	client := meta.(*provider.Context).Client
-	ctx := context.Background()
-	name := d.Get("name").(string)
-	id := sdk.NewAccountObjectIdentifier(name)
-
-	// Is it a Shared Database?
-	if fromShare, ok := d.GetOk("from_share"); ok {
-		account := fromShare.(map[string]interface{})["provider"].(string)
-		share := fromShare.(map[string]interface{})["share"].(string)
-		shareID := sdk.NewExternalObjectIdentifier(sdk.NewAccountIdentifierFromAccountLocator(account), sdk.NewAccountObjectIdentifier(share))
-		opts := &sdk.CreateSharedDatabaseOptions{}
-		if v, ok := d.GetOk("comment"); ok {
-			opts.Comment = sdk.String(v.(string))
-		}
-		err := client.Databases.CreateShared(ctx, id, shareID, opts)
-		if err != nil {
-			return fmt.Errorf("error creating database %v: %w", name, err)
-		}
-		d.SetId(name)
-		return ReadDatabase(d, meta)
-	}
-	// Is it a Secondary Database?
-	if primaryName, ok := d.GetOk("from_replica"); ok {
-		primaryID := sdk.NewExternalObjectIdentifierFromFullyQualifiedName(primaryName.(string))
-		opts := &sdk.CreateSecondaryDatabaseOptions{}
-		if v := d.Get("data_retention_time_in_days"); v.(int) != -1 {
-			opts.DataRetentionTimeInDays = sdk.Int(v.(int))
-		}
-		err := client.Databases.CreateSecondary(ctx, id, primaryID, opts)
-		if err != nil {
-			return fmt.Errorf("error creating database %v: %w", name, err)
-		}
-		d.SetId(name)
-		// todo: add failover_configuration block
-		return ReadDatabase(d, meta)
-	}
 
-	// Otherwise it is a Standard Database
-	opts := sdk.CreateDatabaseOptions{}
-	if v, ok := d.GetOk("comment"); ok {
-		opts.Comment = sdk.String(v.(string))
+	id := sdk.NewAccountObjectIdentifier(d.Get("name").(string))
+
+	dataRetentionTimeInDays,
+		maxDataExtensionTimeInDays,
+		externalVolume,
+		catalog,
+		replaceInvalidCharacters,
+		defaultDDLCollation,
+		storageSerializationPolicy,
+		logLevel,
+		traceLevel,
+		suspendTaskAfterNumFailures,
+		taskAutoRetryAttempts,
+		userTaskManagedInitialWarehouseSize,
+		userTaskTimeoutMs,
+		userTaskMinimumTriggerIntervalInSeconds,
+		quotedIdentifiersIgnoreCase,
+		enableConsoleOutput := GetAllDatabaseParameters(d)
```

Comment:
I don't like this, it's dangerous assignment, let's leave it for now, but we should have a map or SDK object for it, please add a TODO (in the bag)

Reply:
Added